File: CSV2HTML_by_Java/csv2html/Translator.java
package csv2html;

import java.awt.image.BufferedImage;
import java.io.File;
import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.HashSet;
import java.util.List;

/**
 * トランスレータ：CSVファイルをHTMLページへと変換するプログラム。
 */
public class Translator extends Object
{
	/**
	 * CSVに由来するテーブルを記憶するフィールド。
	 */
	private Table inputTable;

	/**
	 * HTMLに由来するテーブルを記憶するフィールド。
	 */
	private Table outputTable;

	/**
	 * 属性リストのクラスを指定したトランスレータのコンストラクタ。
	 * @param classOfAttributes 属性リストのクラス
	 */
	public Translator(Class<? extends Attributes> classOfAttributes)
	{
		super();

		Attributes.flushBaseDirectory();

		try
		{
			Constructor<? extends Attributes> aConstructor = classOfAttributes.getConstructor(String.class);

			this.inputTable = new Table(aConstructor.newInstance("input"));
			this.outputTable = new Table(aConstructor.newInstance("output"));
		}
		catch (NoSuchMethodException |
			   InstantiationException |
			   IllegalAccessException |
			   InvocationTargetException anException) { anException.printStackTrace(); }

		return;
	}

	/**
	 * 在位日数を計算して、それを文字列にして応答する。
	 * @param periodString 在位期間の文字列
	 * @return 在位日数の文字列
	 */
	public String computeNumberOfDays(String periodString)
	{
		// 日付文字列を「〜」で分割
		String[] dates = periodString.split("〜");
		if (dates.length < 2) { return ""; }

		try
		{
			// 日付フォーマットを定義
			java.text.SimpleDateFormat format = new java.text.SimpleDateFormat("yyyy年MM月dd日");

			// 開始日と終了日をパース
			java.util.Date startDate = format.parse(dates[0].trim());

			// 終了日が空の場合は現在日を使用
			java.util.Date endDate;
			if (dates[1].trim().isEmpty())
			{
				endDate = new java.util.Date();
			}
			else
			{
				endDate = format.parse(dates[1].trim());
			}

			// 日数を計算（ミリ秒の差を日数に変換、開始日と終了日の両方を含むため+1）
			long diffInMillis = endDate.getTime() - startDate.getTime();
			long days = diffInMillis / (1000 * 60 * 60 * 24) + 1;

			return String.valueOf(days);
		}
		catch (Exception anException)
		{
			// パースエラーの場合は空文字列を返す
			return "";
		}
	}

	/**
	 * サムネイル画像から画像へ飛ぶためのHTML文字列を作成して、それを応答する。
	 * @param imageString 画像の文字列
	 * @param inputTuple 入力タプル
	 * @param tupleIndex タプルのインデックス
	 * @return サムネイル画像から画像へ飛ぶためのHTML文字列
	 */
	public String computeStringOfImage(String imageString, Tuple inputTuple, int tupleIndex)
	{
		// 入力テーブルの属性を取得
		Attributes inputAttributes = this.inputTable.attributes();

		// サムネイルのインデックスを取得し、サムネイル文字列を取得
		int thumbnailIndex = inputAttributes.indexOfThumbnail();
		String thumbnailString = imageString; // デフォルトは画像と同じ
		if (thumbnailIndex >= 0 && thumbnailIndex < inputTuple.values().size())
		{
			String thumb = inputTuple.values().get(thumbnailIndex);
			if (thumb != null && !thumb.isEmpty())
			{
				thumbnailString = thumb;
			}
		}

		// baseUrlを取得
		String baseUrl = inputAttributes.baseUrl();

		// 画像のフルURL
		String imageUrl = baseUrl + imageString;

		// サムネイルのフルURL
		String thumbnailUrl = baseUrl + thumbnailString;

		// ファイル名を取得（altテキスト用）
		String altText = imageString;
		int lastSlash = imageString.lastIndexOf('/');
		if (lastSlash >= 0 && lastSlash < imageString.length() - 1)
		{
			altText = imageString.substring(lastSlash + 1);
		}

		// HTML文字列を生成: <a href="画像URL"><img src="サムネイルURL" alt="ファイル名" /></a>
		StringBuilder htmlBuilder = new StringBuilder();
		htmlBuilder.append("<a href=\"").append(imageUrl).append("\">");
		htmlBuilder.append("<img src=\"").append(thumbnailUrl).append("\" alt=\"").append(altText).append("\" />");
		htmlBuilder.append("</a>");

		return htmlBuilder.toString();
	}

	/**
	 * CSVファイルをHTMLページへ変換する。
	 */
	public void execute()
	{
		// 必要な情報をダウンロードする。
		Downloader aDownloader = new Downloader(this.inputTable);
		aDownloader.perform();

		// ダウンロードしたCSVファイルを読み込む。
		Reader aReader = new Reader(this.inputTable);
		aReader.perform();

		// CSVに由来するテーブルをHTMLに由来するテーブルへと変換する。
		System.out.println(this.inputTable);
		this.translate();
		System.out.println(this.outputTable);

		// HTMLに由来するテーブルから書き出す。
		Writer aWriter = new Writer(this.outputTable);
		aWriter.perform();

		// ブラウザを立ち上げて閲覧する。
		try
		{
			Attributes attributes = this.outputTable.attributes();
			String fileStringOfHTML = attributes.baseDirectory() + attributes.indexHTML();
			ProcessBuilder aProcessBuilder = new ProcessBuilder("open", "-a", "Safari", fileStringOfHTML);
			aProcessBuilder.start();
		}
		catch (Exception anException) { anException.printStackTrace(); }

		return;
	}

	/**
	 * 属性リストのクラスを受け取って、CSVファイルをHTMLページへと変換するクラスメソッド。
	 * @param classOfAttributes 属性リストのクラス
	 */
	public static void perform(Class<? extends Attributes> classOfAttributes)
	{
		// トランスレータのインスタンスを生成する。
		Translator aTranslator = new Translator(classOfAttributes);
		// トランスレータにCSVファイルをHTMLページへ変換するように依頼する。
		aTranslator.execute();

		return;
	}

	/**
	 * CSVファイルを基にしたテーブルから、HTMLページを基にするテーブルに変換する。
	 */
	public void translate()
	{
		Attributes inputAttributes = this.inputTable.attributes();
		Attributes outputAttributes = this.outputTable.attributes();

		// inputTableの各Tupleを処理
		int tupleIndex = 0;
		for (Tuple inputTuple : this.inputTable.tuples())
		{
			List<String> outputValues = new ArrayList<String>();

			// outputのkeysに合わせて値を設定
			for (int index = 0; index < outputAttributes.size(); index++)
			{
				String key = outputAttributes.keys().get(index);

				if (key.equals("days"))
				{
					// 在位日数を計算
					int periodIndex = inputAttributes.indexOf("period");
					if (periodIndex >= 0 && periodIndex < inputTuple.values().size())
					{
						String periodString = inputTuple.values().get(periodIndex);
						String daysString = this.computeNumberOfDays(periodString);
						outputValues.add(daysString);
					}
					else
					{
						outputValues.add("");
					}
				}
				else if (key.equals("image"))
				{
					// 画像列はHTML文字列に変換
					int imageIndex = inputAttributes.indexOfImage();
					if (imageIndex >= 0 && imageIndex < inputTuple.values().size())
					{
						String imageString = inputTuple.values().get(imageIndex);
						String imageHtml = this.computeStringOfImage(imageString, inputTuple, tupleIndex);
						outputValues.add(imageHtml);
					}
					else
					{
						outputValues.add("");
					}
				}
				else
				{
					// 対応するinputの値を取得
					int inputIndex = inputAttributes.indexOf(key);
					if (inputIndex >= 0 && inputIndex < inputTuple.values().size())
					{
						outputValues.add(inputTuple.values().get(inputIndex));
					}
					else
					{
						outputValues.add("");
					}
				}
			}

			// 新しいTupleを作成してoutputTableに追加
			Tuple outputTuple = new Tuple(outputAttributes, outputValues);
			this.outputTable.add(outputTuple);
			tupleIndex++;
		}

		return;
	}
}

File: CSV2HTML_by_Python/csv2html/translator.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""CSV2HTML：総理大臣と徳川幕府の情報「CSV」からWebページ「HTML」を生成。"""
__author__ = 'AOKI Atsushi'
__version__ = '1.0.7'
__date__ = '2021/01/10 (Created: 2016/01/01)'

import datetime
# import locale
import os
import os.path
import re
import subprocess

# from PIL import Image

from csv2html.downloader import Downloader
from csv2html.io import IO
from csv2html.reader import Reader
from csv2html.table import Table
from csv2html.tuple import Tuple
from csv2html.writer import Writer

# pylint: disable=R0201
# R0201: Method could be a function (no-self-use)

class Translator:
	"""トランスレータ：CSVファイルをHTMLページへと変換するプログラム。"""

	def __init__(self, classOfAttributes):
		"""トランスレータのコンストラクタ。"""

		super().__init__()
		classOfAttributes.flush_base_directory()
		self._input_table = Table('input', classOfAttributes)
		self._output_table = Table('output', classOfAttributes)

	def compute_string_of_days(self, period):
		"""在位日数を計算して、それを文字列にして応答する。"""

		# 在位期間の文字列から在位日数を計算する
		try:
			dates = period.split('〜')
			start_date_str = dates[0]
			end_date_str = dates[1] if len(dates) > 1 else ''

			start_date = datetime.datetime.strptime(start_date_str, '%Y年%m月%d日')
			
			if end_date_str == '' or end_date_str is None:
				end_date = datetime.datetime.now()
			else:
				end_date = datetime.datetime.strptime(end_date_str, '%Y年%m月%d日')

			delta = end_date - start_date
			days = delta.days + 1
			
			return f"{days:,}"
		except Exception:
			return ''

	def compute_string_of_image(self, a_tuple):
		"""サムネイル画像から画像へ飛ぶためのHTML文字列を作成して、それを応答する。"""

		# 画像のHTML文字列を作成する
		
		attributes = a_tuple.attributes()
		keys = attributes.keys()
		values = a_tuple.values()
		
		# 画像とサムネイルのファイル名を取得
		image_filename = ""
		thumbnail_filename = ""
		
		try:
			image_index = keys.index("image")
			if image_index < len(values):
				image_filename = values[image_index]
		except ValueError:
			pass
			
		try:
			thumbnail_index = keys.index("thumbnail")
			if thumbnail_index < len(values):
				thumbnail_filename = values[thumbnail_index]
			else:
				thumbnail_filename = image_filename
		except ValueError:
			thumbnail_filename = image_filename

		if not image_filename:
			return ""

		base_url = attributes.base_url()

		
		image_url = base_url + image_filename
		thumbnail_url = base_url + thumbnail_filename
		
		return f'<a href="{image_url}"><img src="{thumbnail_url}" alt="{image_filename}" /></a>'

	def execute(self):
		"""CSVファイルをHTMLページへと変換する。"""

		# ダウンローダに必要なファイル群をすべてダウンロードしてもらい、
		# 入力となるテーブルを獲得する。
		a_downloader = Downloader(self._input_table)
		a_downloader.perform()

		# ダウンロードしたCSVファイルを読み込む。
		a_reader = Reader(self._input_table)
		a_reader.perform()

		# トランスレータに入力となるテーブルを渡して変換してもらい、
		# 出力となるテーブルを獲得する。
		print(self._input_table)
		self.translate()
		print(self._output_table)

		# ライタに出力となるテーブルを渡して、
		# Webページを作成してもらう。
		a_writer = Writer(self._output_table)
		a_writer.perform()

		# 作成したページをウェブブラウザで閲覧する。
		class_attributes = self._output_table.attributes().__class__
		base_directory = class_attributes.base_directory()
		index_html = class_attributes.index_html()
		a_command = "open -a 'Safari' " + base_directory + os.sep + index_html
		subprocess.getoutput(a_command)

	@classmethod
	def perform(cls, class_attributes):
		"""属性リストのクラスを受け取り、CSVファイルをHTMLページへと変換する。"""

		# トランスレータのインスタンスを生成する。
		a_translator = cls(class_attributes)
		# トランスレータにCSVファイルをHTMLページへ変換するように依頼する。
		a_translator.execute()

	def translate(self):
		"""CSVファイルを基にしたテーブルから、HTMLページを基にするテーブルに変換する。"""

		# 入力テーブルの各タプルを処理して、出力テーブルに追加する
		input_attributes = self._input_table.attributes()
		output_attributes = self._output_table.attributes()
		
		output_keys = output_attributes.keys()
		
		for input_tuple in self._input_table.tuples():
			output_values = []
			for key in output_keys:
				if key == 'days':
					# 在位期間から日数を計算
					try:
						period_index = input_attributes.keys().index('period')
						period = input_tuple.values()[period_index]
						days_string = self.compute_string_of_days(period)
						output_values.append(days_string)
					except ValueError:
						output_values.append('')
				elif key == 'image':
					# 画像HTMLを生成
					image_html = self.compute_string_of_image(input_tuple)
					output_values.append(image_html)
				else:
					# その他のキーはそのままコピー
					try:
						input_index = input_attributes.keys().index(key)
						output_values.append(input_tuple.values()[input_index])
					except ValueError:
						output_values.append('')
			
			# 出力タプルを作成してテーブルに追加
			output_tuple = Tuple(output_attributes, output_values)
			self._output_table.add(output_tuple)
